resources/aws_ec2.go

```diff
@@ -555,6 +491,73 @@ func (obj *AwsEc2Res) longpollWatch() error {
 	}
 }
 
+// watchWaiter waits for an instance to change state and returns the new state.
+func watchWaiter(ctx context.Context, instance *ec2.Instance, c *ec2.EC2) (string, error) {
+	var err error
+	var name string
+
+	// these cases are not permitted.
+	if instance == nil {
+		return "", fmt.Errorf("nil instance")
+	}
+	if aws.StringValue(instance.State.Name) == "" {
+		return "", fmt.Errorf("nil or empty state")
+	}
+
+	// get the instance name.
+	for _, tag := range instance.Tags {
+		if aws.StringValue(tag.Key) == nameKey {
+			name = aws.StringValue(tag.Value)
+		}
+	}
+	// error if we didn't find one.
+	if name == "" {
+		return "", fmt.Errorf("name not found")
+	}
+
+	// instance.InstanceId might be nil in case of terminated instances, which
+	// the api will treat as if it isn't included.
+	waitInput := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{instance.InstanceId},
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(name)},
+			},
+		},
+	}
+	// call on the appropriate waiter based on current state.
+	switch aws.StringValue(instance.State.Name) {
+	case ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping:
+		err = c.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameStopped, ec2.InstanceStateNamePending:
+		err = c.WaitUntilInstanceRunningWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameTerminated:
+		err = c.WaitUntilInstanceExistsWithContext(ctx, waitInput)
+	}
+	if err != nil {
+		aerr, ok := err.(awserr.Error)
+		if !ok {
+			return "", errwrap.Wrapf(err, "error casting awserr")
+		}
+		// ignore these errors
+		if aerr.Code() != request.CanceledErrorCode && aerr.Code() != request.WaiterResourceNotReadyErrorCode {
+			return "", errwrap.Wrapf(err, "internal waiter error")
+		}
+		// if the waiter times out, start again recursively.
+		if aerr.Message() == AwsErrExceededWaitAttempts {
+			return watchWaiter(ctx, instance, c)
```

Comment:
While this might logically make sense, we shouldn't push code that depends on an infinitely deep recursion call stack, since *eventually* we would run out of memory. Not that it's necessarily likely to happen (I didn't measure) but it's logically inappropriate. So we need a different mechanism please.

Reply:
OK. Good thing we didn't go with HASKELL :P 
Do you want me to fix this before you review?